Interference in parallel tests with TestHonestMultiInstance_Agreement

When enabling more test parallelism, I ran into a failing test for multi-instance agreement only when run in parallel with other tests. When `TestHonestMultiInstance_Agreement` is run in isolation, it always passes. When run in parallel with all other tests, the third iteration always fails with too many rounds. Parallelism was obtained with `t.Parallel()` in both `TestHonestMultiInstance_Agreement` and inside the nested `t.Run()`.

My IDE is running `go test -json -parallel=8 -count=1 ./... -run ./...`, and I replicate the failure with this at CLI. However, removing the `-json` flag causes the tests to pass at CLI 😕.

```
=== RUN   TestHonestMultiInstance_Agreement/async_3
=== PAUSE TestHonestMultiInstance_Agreement/async_3
=== CONT  TestHonestMultiInstance_Agreement/async_3
    multi_instance_test.go:113: 
        	Error Trace:	/Users/me/pl/src/filecoin-project/go-f3/test/multi_instance_test.go:113
        	            				/Users/me/pl/src/filecoin-project/go-f3/test/multi_instance_test.go:46
        	Error:      	Received unexpected error:
        	            	reached maximum number of 10 rounds
        	Test:       	TestHonestMultiInstance_Agreement/async_3
        	Messages:   	P0{593}, round 10, phase COMMIT
        	            	P1{593}, round 11, phase CONVERGE
        	            	P2{593}, round 10, phase COMMIT
--- FAIL: TestHonestMultiInstance_Agreement/async_3 (0.25s)
```

--- a/sim/cmd/f3sim/f3sim.go
+++ b/sim/cmd/f3sim/f3sim.go
@@ -31,7 +31,10 @@ func main() {
 		fmt.Printf("Iteration %d: seed=%d, mean=%f\n", i, seed, *latencyMean)
 
 		options := []sim.Option{
-			sim.WithLatencyModel(latency.NewLogNormal(*latencySeed, time.Duration(*latencyMean*float64(time.Second)))),
+			sim.WithLatencyModeler(func() (latency.Model, error) {
+				mean := time.Duration(*latencyMean * float64(time.Second))
+				return latency.NewLogNormal(*latencySeed, mean), nil
+			}),
 			sim.WithECEpochDuration(30 * time.Second),
 			sim.WithECStabilisationDelay(0),
 			sim.AddHonestParticipants(

--- a/sim/latency/latency.go
+++ b/sim/latency/latency.go
@@ -6,6 +6,9 @@ import (
 	"github.com/filecoin-project/go-f3/gpbft"
 )
 
+// Modeler instantiates a new latency Model.
+type Modeler func() (Model, error)
+
 // Model represents a latency model of cross participant communication. The model
 // offers the ability for implementation of varying latency across a simulation,
 // as well as specialised latency across specific participants.

--- a/sim/options.go
+++ b/sim/options.go
@@ -93,10 +93,11 @@ func WithSigningBackend(sb signing.Backend) Option {
 	}
 }
 
-func WithLatencyModel(lm latency.Model) Option {
+func WithLatencyModeler(lm latency.Modeler) Option {
 	return func(o *options) error {
-		o.latencyModel = lm
-		return nil
+		var err error
+		o.latencyModel, err = lm()
+		return err
 	}
 }
 

--- a/test/constants_test.go
+++ b/test/constants_test.go
@@ -34,7 +34,7 @@ var (
 
 func syncOptions(o ...sim.Option) []sim.Option {
 	return append(o,
-		sim.WithLatencyModel(latency.None),
+		sim.WithLatencyModeler(func() (latency.Model, error) { return latency.None, nil }),
 		sim.WithECEpochDuration(EcEpochDuration),
 		sim.WitECStabilisationDelay(EcStabilisationDelay),
 		sim.WithGpbftOptions(testGpbftOptions...),
@@ -43,7 +43,9 @@ func syncOptions(o ...sim.Option) []sim.Option {
 
 func asyncOptions(latencySeed int, o ...sim.Option) []sim.Option {
 	return append(o,
-		sim.WithLatencyModel(latency.NewLogNormal(int64(latencySeed), latencyAsync)),
+		sim.WithLatencyModeler(func() (latency.Model, error) {
+			return latency.NewLogNormal(int64(latencySeed), latencyAsync), nil
+		}),
 		sim.WithECEpochDuration(EcEpochDuration),
 		sim.WitECStabilisationDelay(EcStabilisationDelay),
 		sim.WithGpbftOptions(testGpbftOptions...),

--- a/test/multi_instance_test.go
+++ b/test/multi_instance_test.go
@@ -11,8 +11,7 @@ import (
 
 // TestHonestMultiInstance_Agreement tests for multiple chained instances of the protocol with no adversaries.
 func TestHonestMultiInstance_Agreement(t *testing.T) {
-	// TODO: t.Parallel() here after https://github.com/filecoin-project/builtin-actors/issues/1541
-	//t.Parallel()
+	t.Parallel()
 	const (
 		instanceCount  = 4000
 		testRNGSeed    = 8965130
@@ -101,8 +100,7 @@ func FuzzHonestMultiInstance_AsyncAgreement(f *testing.F) {
 }
 
 func multiAgreementTest(t *testing.T, seed int, honestCount int, instanceCount uint64, maxRounds uint64, opts ...sim.Option) {
-	// TODO: t.Parallel() here after https://github.com/filecoin-project/builtin-actors/issues/1541
-	//t.Parallel()
+	t.Parallel()
 	rng := rand.New(rand.NewSource(int64(seed)))
 	sm, err := sim.NewSimulation(append(opts,
 		sim.AddHonestParticipants(

--- a/test/spam_test.go
+++ b/test/spam_test.go
@@ -44,7 +44,9 @@ func TestSpamAdversary(t *testing.T) {
 			t.Run(name, func(t *testing.T) {
 				ecChainGenerator := sim.NewUniformECChainGenerator(651651, 1, 10)
 				sm, err := sim.NewSimulation(
-					sim.WithLatencyModel(latency.NewLogNormal(455454, time.Second)),
+					sim.WithLatencyModeler(func() (latency.Model, error) {
+						return latency.NewLogNormal(455454, time.Second), nil
+					}),
 					sim.WithECEpochDuration(EcEpochDuration),
 					sim.WitECStabilisationDelay(EcStabilisationDelay),
 					sim.WithGpbftOptions(testGpbftOptions...),

--- a/test/withhold_test.go
+++ b/test/withhold_test.go
@@ -37,14 +37,16 @@ func TestWitholdCommitAdversary(t *testing.T) {
 	for _, test := range tests {
 		test := test
 		t.Run(test.name, func(t *testing.T) {
-			nearSynchrony := latency.NewLogNormal(1413, 10*time.Millisecond)
+			nearSynchrony := func() (latency.Model, error) {
+				return latency.NewLogNormal(1413, 10*time.Millisecond), nil
+			}
 			tsg := sim.NewTipSetGenerator(tipSetGeneratorSeed)
 			baseChain := generateECChain(t, tsg)
 			a := baseChain.Extend(tsg.Sample())
 			b := baseChain.Extend(tsg.Sample())
 			victims := []gpbft.ActorID{0, 1, 2, 3}
 			sm, err := sim.NewSimulation(
-				sim.WithLatencyModel(nearSynchrony),
+				sim.WithLatencyModeler(nearSynchrony),
 				sim.WithECEpochDuration(EcEpochDuration),
 				sim.WitECStabilisationDelay(EcStabilisationDelay),
 				sim.WithGpbftOptions(testGpbftOptions...),

